src/slurm_plugin/clustermgtd.py

```diff
@@ -904,6 +1042,135 @@ def _terminate_orphaned_instances(self, cluster_instances, ips_used_by_slurm):
                 instances_to_terminate, terminate_batch_size=self._config.terminate_max_batch_size
             )
 
+    def _is_enter_protected_mode_required(self, active_partitions_protected_failure_count_map):
+        """Enter protected mode if partition protected count reaches threshod and procted mode enabled."""
+        if self._partitions_protected_failure_count_map:
+            logging.info(f"Partitions_protected_failure_count_map is {self._partitions_protected_failure_count_map}")
+            return any(
+                active_partitions_protected_failure_count_map.get(p) >= self._config.protected_failure_count
+                for p in active_partitions_protected_failure_count_map
+            )
+        # If self._partitions_protected_failure_count_map is None, protected mode is disabled.
+        return False
+
+    def _enter_protected_mode(self, slurm_nodes, active_partitions_protected_failure_count_map):
+        """Entering protected mode if no active running job in queue."""
+        (
+            partitions_to_disable,
+            bootstrap_failure_partitions_have_jobs,
+        ) = self._get_partitions_to_disable_and_bootstrap_failure_partitions_have_jobs(
+            slurm_nodes, active_partitions_protected_failure_count_map
+        )
+        if bootstrap_failure_partitions_have_jobs:
+            log.info(
+                f"Bootstrap failure partitions {bootstrap_failure_partitions_have_jobs} currently have jobs running, "
+                "not disabling them."
+            )
+        if partitions_to_disable:
+            # Change compute fleet status to protected
+            if not self._compute_fleet_status == ComputeFleetStatus.PROTECTED:
+                log.info("Put Cluster into protected mode")
+                self._update_compute_fleet_status(ComputeFleetStatus.PROTECTED)
+            # Place partitions into inactive
+            log.info(f"Placing bootstrap failure partitions to INACTIVE: {partitions_to_disable}")
+            update_partitions(partitions_to_disable, PartitionStatus.INACTIVE)
+        else:
+            log.info("Not entering protected mode since active are jobs running in bootstrap failure partitions")
+
+    def _find_online_slurm_nodes(self, slurm_nodes, private_ip_to_instance_map):
+        """Find slurm nodes online. Consider not DOWN static node and running job dynamic node as online nodes."""
+        online_nodes = []
+        for node in slurm_nodes:
+            if node.is_static and not node.is_down():
+                online_nodes.append(node)
+            else:
+                if self._is_node_healthy(node, private_ip_to_instance_map) and node.is_running_job():
```

Comment:
Why do we need this `is_running_job` logic? What if I submit a job and cancel the job later, but the dynamic node is up and running; should I consider this node as an online node?